feature: 创建UserProjectContact的model以及其的增删改查

YichenWu11 opened the pull request:
创建UserProjectContact的model以及其的增删改查

YichenWu11 committed:
feature: 创建UserProjectContact的model以及其的增删改查

wheel-w commented on project/urls.py:
```diff
@@ -17,5 +17,6 @@
 
 router = DefaultRouter()
 router.register(r"project", views.ProjectViewSet)
+router.register(r"userproject-contact", views.UserProjectContactViewSet)
```
project-user吧

wheel-w commented on project/views.py:
```diff
@@ -13,10 +13,15 @@
+
+class UserProjectContactViewSet(viewsets.ModelViewSet):
+    queryset = UserProjectContact.objects.all()
+    serializer_class = UserProjectContactSerializer
```
1. 这块要可以批量导入用户，传入用户id列表
2. 返回给前端的数据不能只有id，返回给前端的用户列表要有用户对应的属性username name等等

YichenWu11 replied:
修改后的get请求返回的数据
![image](https://user-images.githubusercontent.com/92995653/163138654-aba9c8f9-ac8c-4db9-a371-000de9fdbf37.png)

wheel-w commented on project/serializer.py:
```diff
@@ -13,10 +13,25 @@
+        fields = "__all__"
+
+    # 校验传来的项目id和学生id是否存在
+    def validate(self, attrs):
```
这块只需要校验是否为整数就行
获取项目下用户列表的时候，用户不存在的不会显示的

YichenWu11 replied:
这样的话就不需要在这里进行校验了 model里的两个id都是IntegerField，serializer反序列化时在validate之前会先根据数据在model里的字段类型进行类型校验吧

YichenWu11 committed:
feature: 完善接口，增加导出用户名单

feature: 修改接口

Update serializer.py

Update views.py

minor: 修改接口 (#21)

Delete 0004_userprojectcontact.py

Delete 0005_auto_20220414_2253.py

feature: 修改project-user的接口

YichenWu11 committed:
Merge branch 'delopy' into delopy

wheel-w commented on project/views.py:
```diff
@@ -62,3 +70,94 @@ def partial_update(self, request, *args, **kwargs):
+    def retrieve(self, request, *args, **kwargs):
+        project_id = kwargs["project_id"]
+        data = UserProjectContact.objects.filter(project_id=project_id)
+        serializer = super().get_serializer(data, many=True)
```
获取项目下的用户信息，需要返回给前端同学用户属性列表

YichenWu11 replied:
这里返回给前端的就是用户的属性列表
![image](https://user-images.githubusercontent.com/92995653/163505627-262117bb-4b0c-43b2-a374-3b84eb108c17.png)

wheel-w commented on project/views.py:
```diff
@@ -62,3 +70,94 @@ def partial_update(self, request, *args, **kwargs):
+        user_id = request.data["user_id"]
+        data = UserProjectContact.objects.create(project_id=project_id, user_id=user_id)
+        serializer = UserProjectContactSerializer(data)
+        return Response(serializer.data)
```
创建返回创建成功就行

wheel-w commented on project/views.py:
```diff
@@ -62,3 +70,94 @@ def partial_update(self, request, *args, **kwargs):
+                UserProjectContact(project_id=project_id, user_id=user_id)
+            )
+        data = UserProjectContact.objects.bulk_create(queryset_list)
+        serializer = UserProjectContactSerializer(data, many=True)
```
批量导入，返回成功和失败就行，前端会在请求成功执行，调用get获取用户列表的

wheel-w commented on project/views.py:
```diff
@@ -62,3 +70,94 @@ def partial_update(self, request, *args, **kwargs):
+        project_id = kwargs["project_id"]
+        title = f"{Project.objects.get(id=project_id).name}用户名单"
+        user_id_list = []
+        user_ids = UserProjectContact.objects.filter(project_id=project_id)
```
使用value_list获取用户id吧

wheel-w commented on project/views.py:
```diff
@@ -62,3 +70,94 @@ def partial_update(self, request, *args, **kwargs):
+        for u_i in user_ids:
+            user_id_list.append(u_i.user_id)
+        users = User.objects.filter(id__in=user_id_list)
+        obj = {"MALE": "男", "FEMALE": "女"}
```
这个引用user model里面定义的枚举

wheel-w commented on project/views.py:
```diff
@@ -62,3 +70,94 @@ def partial_update(self, request, *args, **kwargs):
+            queryset_list.append(
+                UserProjectContact(project_id=project_id, user_id=user_id)
+            )
+        data = UserProjectContact.objects.bulk_create(queryset_list)
```
批量创建要考虑去重，只添加没有的成员

YichenWu11 committed:
minor: 修改创建的返回信息，增加去重

YichenWu11 committed:
Merge remote-tracking branch 'origin/delopy' into delopy

YichenWu11 committed:
Merge branch 'wheel-w:delopy' into delopy

YichenWu11 committed:
Merge remote-tracking branch 'origin/delopy' into delopy

YichenWu11 committed:
feature: 增加参数校验

YichenWu11 committed:
minor: 修改response

YichenWu11 committed:
feature: 获取用户列表增加分页

YichenWu11 committed:
feature: 增加分页

wheel-w commented on project/views.py:
```diff
@@ -62,3 +78,114 @@ def partial_update(self, request, *args, **kwargs):
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+        serializer = self.get_serializer(data, many=True)
```
这样会导致n+1查询，返回值不一定需要序列化
分完页，自己查询下用户信息组装下数据

wheel-w commented on project/views.py:
```diff
@@ -62,3 +78,114 @@ def partial_update(self, request, *args, **kwargs):
+        user_id_list = [
+            val
+            for val in users_id
+            if not exist_contacts.filter(project_id=project_id, user_id=val).exists()
```
不要在循环里再进行数据库操作
value_list对比就行了

YichenWu11 committed:
minor: 修改list方法

wheel-w commented on project/views.py:
```diff
@@ -62,3 +77,111 @@ def partial_update(self, request, *args, **kwargs):
+        exist_contacts = UserProjectContact.objects.filter(
+            project_id=project_id
+        ).values_list("user_id", flat=True)
+        exist_users = User.objects.filter(id__in=users_id).values_list("id", flat=True)
```
这个就不用查询了，直接根据users_id对比就行了

wheel-w commented on project/views.py:
```diff
@@ -62,3 +77,111 @@ def partial_update(self, request, *args, **kwargs):
+        project_id = kwargs["project_id"]
+        if not Project.objects.filter(id=project_id).exists():
+            return Response(f"id为{project_id}的项目不存在", exception=True)
+        users_id = request.data["user_id_list"]
```
user_id_list= request.data["user_id_list"]

wheel-w commented on project/views.py:
```diff
@@ -62,3 +77,111 @@ def partial_update(self, request, *args, **kwargs):
+        ).values_list("user_id", flat=True)
+        exist_users = User.objects.filter(id__in=users_id).values_list("id", flat=True)
+        # 去重
+        user_id_list = [val for val in exist_users if val not in exist_contacts]
```
去重用set集合特性

wheel-w commented on project/views.py:
```diff
@@ -62,3 +77,111 @@ def partial_update(self, request, *args, **kwargs):
+    )
+    def destroy(self, request, *args, **kwargs):
+        project_id = kwargs["project_id"]
+        if not Project.objects.filter(id=project_id).exists():
```
上面这个校验可以不要，项目id不存在，他删除了0条数据

wheel-w commented on project/views.py:
```diff
@@ -62,3 +77,111 @@ def partial_update(self, request, *args, **kwargs):
+        # 查询记录数据
+        records = []
+        project_id = kwargs["project_id"]
+        if not Project.objects.filter(id=project_id).exists():
```
这个也不要了，id不存在就查不到
create的时候校验，是避免创建不存在课程的关系

YichenWu11 committed:
minor: 使用set去重

YichenWu11 committed:
minor: 修改序列化器
